### PR TITLE
JWT authentication

### PR DESCRIPTION
**Pull request by achouippe**
- Add JWT authentication to the public and internal apis,
- Rework the contracts of the public and internal apis.
- Rework the way to configure JWT authentication settings per environment.

### CODE REVIEW

File: lib/neurow/jwt_auth_plug.ex [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  defp forbidden(conn, error_code, error_message) do
+    {:ok, response} =
+      Jason.encode(%{
+        errors: [
+          %{error_code: error_code, error_message: error_message}
+        ]
+      })
+
+    conn
```
Comment (achouippe) [inline lib/neurow/jwt_auth_plug.ex:131]: I trace very detailed error code for the sake of debuggability, but maybe I should add a condition to not return this very detailed error causes in production.

File: lib/neurow/jwt_auth_plug.ex [hunk 2]
```diff
@@ -0,0 +1,136 @@
+defmodule Neurow.JwtAuthPlug do
+  require Logger
+
+  import Plug.Conn
+
+  def init(options), do: options
+
+  def call(conn, options) do
+    case jwt_token_from_request(conn) do
```
Comment (achouippe) [inline lib/neurow/jwt_auth_plug.ex:9]: Try to use what seems to be idiomatic of Elixir, but I wonder if I should not use a try / rescue instead.

File: lib/neurow/public_api.ex [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    max_lifetime: 60 * 2,
+    audience: "public_api"
+  )
+
+  plug(:match)
+  plug(:dispatch)
+
+  get "v1/subscribe" do
+    case conn.assigns[:jwt_payload] do
+      %{"iss" => issuer, "sub" => sub} ->
```
Comment (achouippe) [inline lib/neurow/public_api.ex:20]: The idea here is to use the issuer as a way to isolate multiple systems that use neurow for each other.

File: config/runtime.exs [hunk 4]
```diff
@@ -13,3 +13,13 @@ config :neurow, sse_timeout: String.to_integer(System.get_env("SSE_TIMEOUT") ||
 
 config :neurow, ssl_keyfile: System.get_env("SSL_KEYFILE")
 config :neurow, ssl_certfile: System.get_env("SSL_CERTFILE")
+
+config :neurow,
```
Comment (bpaquet) [inline config/runtime.exs:17]: How do we inject this in production? Can we define a json format?

Comment (achouippe) [reply config/runtime.exs:17]: I will update the code so we have a `dev.exs` that directly contains test keys, and `prod.exs` that parse an environment variable.

File: lib/neurow/public_api.ex [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        topic = "#{issuer}-#{sub}"
```
[the remark this answers is not in the document]
Comment (achouippe) [reply lib/neurow/public_api.ex:21]: I considered it, but we need to discuss it. Where to we set the topic & sub topic part ? The sub will eventually contains something like `user:123`. By generating a topic name like `issuer-user:123` the sub topic is used for the record id.

But the purpose of topic / subtopic in the phoenix pubsub is not totally clear for me for now.

**Review by bpaquet — APPROVED** [review]

**Pull request merged**